<a href="https://colab.research.google.com/github/maggiecrowner/labs_fork/blob/main/Wrangling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Wrangling Lab - Maggie Crowner

**Question 2**

*Part 1*

In [11]:
import pandas as pd
! git clone https://github.com/DS3001/wrangling

data = pd.read_csv('/content/wrangling/assignment/data/airbnb_hw.csv', low_memory=False)
price = data['Price']
# print(list(price))
price = price.str.replace(',','')
price = pd.to_numeric(price)
print(list(price))
data['Price'] = price

missing = pd.isnull(price)
print('Total missing: ', sum(missing))

fatal: destination path 'wrangling' already exists and is not an empty directory.
[145, 37, 28, 199, 549, 149, 250, 90, 270, 290, 170, 59, 49, 68, 285, 75, 145, 100, 150, 700, 125, 175, 90, 40, 89, 95, 99, 149, 499, 99, 99, 99, 125, 170, 120, 79, 125, 110, 89, 180, 180, 143, 230, 350, 135, 85, 85, 85, 60, 70, 55, 44, 200, 175, 250, 165, 115, 145, 200, 74, 74, 84, 125, 129, 50, 50, 185, 185, 150, 80, 190, 190, 190, 175, 190, 140, 140, 75, 45, 45, 135, 100, 145, 135, 200, 65, 165, 145, 135, 225, 250, 250, 600, 145, 149, 150, 200, 120, 150, 175, 109, 1990, 80, 73, 135, 89, 89, 250, 99, 250, 250, 250, 240, 115, 72, 165, 90, 105, 155, 160, 150, 89, 99, 42, 42, 132, 117, 135, 125, 200, 250, 295, 280, 75, 150, 125, 230, 110, 159, 160, 107, 50, 69, 110, 175, 125, 95, 95, 135, 100, 200, 239, 220, 199, 399, 130, 70, 110, 160, 135, 350, 60, 375, 225, 165, 585, 40, 85, 225, 110, 125, 275, 240, 175, 130, 90, 130, 130, 130, 130, 120, 139, 260, 35, 185, 175, 150, 120, 150, 175, 133, 375, 150, 150, 15

Here I cleaned the Price variable. I noticed that any numbers over 1,000 included a comma, which made it a string variable and not useful for calculation. I removed the commas, and converted it to a numeric variable, and then put the cleaned variable back into the data set.

There are 0 missing data points in the Price variable.

*Part 2*

In [ ]:
data = pd.read_csv('/content/wrangling/data/sharks.csv', low_memory=False)
type = data['Type']
# print(type.value_counts())
type = type.replace(['Sea Disaster', 'Boat', 'Boating', 'Boatomg'], 'Watercraft')
type = type.replace(['Invalid', 'Questionable', 'Unconfirmed', 'Unverified', 'Under investigation'], 'Unclear')
print(type.value_counts())
data['Type'] = type

Type
Unprovoked    4716
Provoked       593
Watercraft     583
Unclear        565
Name: count, dtype: int64


Here I cleaned the Type variable. I noticed that a lot of types documented were essentially synonyms, and therefore there were more categories in this variable than were necessary. I combined all water/boat-related types into the "Watercraft" category, and I combined all unknown/questionable types into an "Unclear" category. I then put the cleaned variable back into the data set.

*Part 3*

In [14]:
import numpy as np

url = 'http://www.vcsc.virginia.gov/pretrialdataproject/October%202017%20Cohort_Virginia%20Pretrial%20Data%20Project_Deidentified%20FINAL%20Update_10272021.csv'
data = pd.read_csv(url,low_memory=False)
released = data['WhetherDefendantWasReleasedPretrial']
# print(released.value_counts())
released = released.replace(9,np.nan)
print(released.value_counts())
data['WhetherDefendantWasReleasedPretrial'] = released

WhetherDefendantWasReleasedPretrial
1.0    19154
0.0     3801
Name: count, dtype: int64


In the codebook for this data set, we can find that 1 corresponds to 'Released during pre-trial period', 0 corresponds to 'Detained entire pre-trial period', and 9 corresponds to 'Unknown'. So, here I replaced all of the 9s with np.nan since they are missing values, and then put the cleaned variable back into the data set.

*Part 4*

In [28]:
sentence = data['ImposedSentenceAllChargeInContactEvent']
s_type = data['SentenceTypeAllChargesAtConvictionInContactEvent']
# print(sentence.value_counts())
# print(s_type.value_counts())
sentence = pd.to_numeric(sentence, errors='coerce')
print('Total missing data points before cleaning: ', sum(pd.isnull(sentence)))
sentence = sentence.mask(s_type == 4, 0)
sentence = sentence.mask(s_type == 9, np.nan)
print('Total missing data points: ', sum(pd.isnull(sentence)))
data['ImposedSentenceAllChargeInContactEvent'] = sentence

Total missing data points before cleaning:  9053
Total missing data points:  274


Here I cleaned the ImposedSentenceAllChargeInContactEvent variable. From the codebook, you can see that there are actually only 3 types of sentences, corresponding to 0, 1, and 2 in the SentenceTypeAllChargesAtConvictionInContactEvent variable. So, type '4' in that variable should not have any events, and we already know that '9' corresponds to unknown data. So, I changed the 4s to be 0, and the 9s to be np.nan. This brought my amount of missing data points from 9053 to just 274. I then put the cleaned variable back into the data set.

**Question 1**

1. This paper will discuss data tidying. Tidy data has a specific structure and is easy to work with, clean, and visualize, so there are many advantages of using it.

2. The tidy data standard is meant to make data analysis easier by removing the confusing logistical aspect of how to read the data's structure.

3. "Messy" datasets could be cleaned for hours and still be messy because there is no standard format, so each dataset may have to be read totally differently. It is hard to define observations and variables because there are many things you have to consider, such as the population, generalizability, how the data was collected, and how a category is defined. Without a pretty comprehensive data description/dictionary, values could mean literally nothing with no context. The observation could be a specific time, specific person, etc., but how does that fit into the value?

4. A value is described as the actual data points in the data set that you would use for calculation and analysis. An observation is the person, time, thing, or place that is being measured for all variables. A variable is the measurement being checked for all observations, such as temperature or length, for example.

5. Tidy data contains headings for each column which correspond to a variable, headings for each row which correspond to an observation, and values in each spot that corresponds to that particular observation and variable, to form a table.

6. The first most common problem is when column headers are values and not variable names, which means that the "values" in the table are actually counts of how many observations fit into those values. The second most common problem is when multiple variables are stored in one column to attempt to group them prematurely. The third most common problem is when variables are stored in rows and columns. The fourth most common problem is when multiple types of observational units are stored in the same table, which creates a lot of duplicates. The fifth most common problem is when a single observation is stored in multiple tables when they could be combined.

7. Table 11 has a lot of missing data points and is way too long, due to there being an excess of splitting up of variables. This would make calculations difficult. Table 12 shows how you can create just one table which displays all of the data in a more compact way.

8. The chicken and the egg problem with tidy data is that in order for you to use the tidy tools, you may want to use the tidy tools in order to convert the data to a tidy format, but you cannot use those tools until it is already in that format. Wickham hopes that in the future, people will develop better tools and strategies than the tidy data framework that is more foolproof. He also hopes people will develop better frameworks for other tasks, such as parsing data, identifying missing values, matching similar values, etc.